Get space label from Build object and pass to fabric8-maven-plugin

ebaron opened the pull request:
Recently, Spring Boot boosters stopped working with our "Deployments" tab: https://github.com/openshiftio/openshift.io/issues/2360. 

The root cause of the issue is that the Spring Boot boosters moved their fabric8-maven-plugin execution to a new parent project. When creating a new project, the launcher would look for the fabric8-maven-plugin execution in the project's pom and add a label containing the name of the project's space. If fabric8-maven-plugin is missing from the project's pom file, no space label is added to the generated OpenShift resources. There may also be a problem if the project is shared between more than one space, where the launcher would overwrite the previous space name in the FMP configuration.

In search of a more reliable way to set the space name on objects created by our builds, I'd like to propose this PR which leverages information already available in the build. This involves adding a method in Utils to retrieve the space label from the Build object corresponding to the running Jenkins pipeline. This label is then passed as a property to FMP, which boosters (or parents of boosters) can use in their configuration. Such as:
```
 <resources>
   <labels>
     <all>
       <space>${fabric8.space}</space>
     </all>
   </labels>
 </resources>
```

ebaron commented:
cc @chmouel

gastaldi commented:
cc @cmoulliard @metacosm

metacosm commented:
I don't have an OSIO account and have never tried it so, unfortunately, I have no idea what you guys are talking about… :(

jaseemabid commented:
I've nothing meaningful to add to this discussion but I'll go though an end to end test of one or 2 quick start apps before merging this though.

rupalibehera commented:
@ebaron, the patch looks good, I would appreciate if you could test it as mentioned here https://github.com/fabric8io/fabric8-pipeline-library/wiki/FAQ#2-how-do-you-test-the-pipeline-code- on OSIO just point the import statement to your branch in Jenkinsfile.

I tried testing it, but for some reason I am not able to create a quickstart.

ebaron commented:
@rupalibehera Hi, I updated the PR to use a property name suitable for a fabric8-maven-plugin enricher. I have an enricher counterpart for this PR that will allow FMP to apply the space labels without any special configuration.

Since this uses a new Kubernetes client method (ObjectMeta.getLabels), I had to add this method to the whitelist manually and redeploy Jenkins. We will need to make this part of the default config before we can merge this PR: https://github.com/ebaron/fabric8-tenant-jenkins/commit/077a22900bc3a31bdc54a9dc7af4ef405ff4d884. Is fabric8-tenant-jenkins the right place for the PR?

Once the whitelist was updated, the Jenkins build worked as expected.
```
[...]
[Pipeline] echo
Looking for matching Build myotherspace-3
[Pipeline] echo
looking for myotherspace-3 in namespace ebaron1
[Pipeline] echo
Found label for space: 'myotherspace', in Build myotherspace-3
[Pipeline] sh
[ebaron_myotherspace_master-RKPE7GBT34ZE7Y26CCBJD5AXYYVMP23K62IJZMTFQX57YBSTVDZA] Running shell script
+ mvn clean -B -e -U deploy -Dmaven.test.skip=false -Dfabric8.enricher.fmp-space-label.space=myotherspace -P openshift
Picked up _JAVA_OPTIONS: -Duser.home=/root/ -XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap -Dsun.zip.disableMemoryMapping=true -XX:+UseParallelGC -XX:MinHeapFreeRatio=5 -XX:MaxHeapFreeRatio=10 -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -Xms10m -Xmx192m
[...]
```

rupalibehera commented:
@ebaron , your changes for fabric8-tenant Jenkins looks good, that is the file you need to modify. please raise a PR there so we can get that in and then this change. Both the PR's looks good.

rupalibehera commented:
@jaseemabid , when we have a PR from @ebaron for fabric8-tenant Jenkins, please merge that and also please monitor the CD pipeline https://github.com/fabric8io/fabric8-build-team/wiki/Jenkins-Images#3-tenant-jenkins-fabric8-tenant-jenkins. And when the fabric8-tenant changes goes in please merge this PR too. 
To get these changes to PROD we need to raise an issue to platform team for the same, like this https://github.com/openshiftio/openshift.io/issues/2670

jaseemabid commented on vars/mavenCanaryRelease.groovy:
```diff
@@ -7,6 +7,9 @@ import org.apache.maven.model.Profile
 @Field final String FMP_STABLE_VERSION = "3.5.38"
 
+// First version of FMP to include 'osio' profile
+@Field final String FMP_OSIO_MIN_VERSION = "3.5.40"
```
@piyush1594 says "3.5.40" is stable now.

piyush1594 replied:
Yes it is a stable version and also maintaining the two FMP versions here is not a good thing

ebaron replied:
@jaseemabid @piyush1594 If the auto-update of FMP is disabled, or does not update for any reason, this would still allow us to apply space labels for FMP versions >= 3.5.40 but less than FMP_STABLE_VERSION.

FMP_OSIO_MIN_VERSION should never need to change. It should only used to check that the `osio` FMP profile exists. I could reduce the scope of this variable to the function it is used in. What do you think?

piyush1594 replied:
@ebaron Got your point. you want this to check that profile extend feature is there or not that is available from FMP 3.5.40. Cool then.

Please also change the FMP Stable version to 3.5.40, changes in boosters are also in the process so.

ebaron replied:
@piyush1594 Right, 3.5.40 brings the profile extends functionality, the `osio` profile, and the enricher for applying space labels. All of these things are required for this PR.

For better clarity, I have renamed the FMP_OSIO_MIN_VERSION constant to FMP_OSIO_PROFILE_MIN_VERSION, and added a comment specifying that it should not be changed.

I would gladly update the stable version to 3.5.40, but I encountered a slight problem. We also need to change the reference versions of booster-parent and spring-boot-booster-parent used by comparisons in `patchPluginBasedOnParent` and `patchProfileBasedOnParent`. The booster-parent update has been released as version 23, but there is no such release yet for spring-boot-booster-parent. I'm not sure if this will be released as 1.5.13-3 or 1.5.14. Any idea?

https://github.com/fabric8io/fabric8-pipeline-library/blob/205dcb720aea989c647b145262c461ef9f09b05c/vars/mavenCanaryRelease.groovy#L190-L194

https://github.com/fabric8io/fabric8-pipeline-library/blob/205dcb720aea989c647b145262c461ef9f09b05c/vars/mavenCanaryRelease.groovy#L212-L216

piyush1594 commented on vars/mavenCanaryRelease.groovy:
```diff
@@ -7,6 +7,9 @@ import org.apache.maven.model.Profile
 
 @Field final String FMP_STABLE_VERSION = "3.5.38"
```
yes, @rupalibehera I think we should upgrade this also to FMP 3.5.40, so that it will also help us in https://github.com/openshiftio/openshift.io/issues/3058

ebaron commented:
@rupalibehera @piyush1594 I have changed `FMP_STABLE_VERSION` to 3.5.40, and also updated the versions of the parent boosters we use to check whether to patch the project's FMP version. These should be the first version of the parent that uses the current stable FMP version.

It appears that spring-boot-booster-parent is deprecated and Spring Boot boosters now inherit from booster-parent instead. Since no release of spring-boot-booster-parent was done that includes FMP 3.5.40, I have removed the version comparison for this case so we always patch the version of FMP.
